src/libstd/sys/unix/condvar.rs

```diff
@@ -60,12 +60,17 @@ impl Condvar {
         let r = ffi::gettimeofday(&mut sys_now, ptr::null_mut());
         debug_assert_eq!(r, 0);
 
+        let nsec = dur.extra_nanos() as libc::c_long +
+                   (sys_now.tv_usec * 1000) as libc::c_long;
+        let extra = (nsec / 1_000_000_000) as libc::time_t;
+        let nsec = nsec % 1_000_000_000;
         let seconds = dur.secs() as libc::time_t;
-        let timeout = match sys_now.tv_sec.checked_add(seconds) {
+
+        let timeout = match (sys_now.tv_sec + extra).checked_add(seconds) {
```

Comment:
I figured that `gettimeofday` won't return a value that is 1 away from overflowing, but we don't control `seconds + sys_now.tv_sec` so that still needs to be checked.